File: cdm/csharp/patient/actions/SETensionPneumothorax.cs
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

public class SETensionPneumothorax : SEPatientAction
{
    protected eGate type;
    protected eSide side;
    protected SEScalar0To1 severity;

    public SETensionPneumothorax()
    {
        type = eGate.NullGate;
        side = eSide.NullSide;
        severity = null;
    }

    public override void Clear()
    {
        base.Clear();
        type = eGate.NullGate;
        side = eSide.NullSide;
        if (severity != null)
            severity.Invalidate();
    }

    public override bool IsValid()
    {
        return HasType() && HasSide() && HasSeverity();
    }

    public new eGate GetType()
    {
        return type;
    }
    public void SetType(eGate type)
    {
        this.type = type;
    }
    public bool HasType()
    {
        return type == eGate.NullGate ? false : true;
    }

    public eSide GetSide()
    {
        return side;
    }
    public void SetSide(eSide side)
    {
        this.side = side;
    }
    public bool HasSide()
    {
        return side == eSide.NullSide ? false : true;
    }

    public bool HasSeverity()
    {
        return severity == null ? false : severity.IsValid();
    }
    public SEScalar0To1 GetSeverity()
    {
        if (severity == null)
            severity = new SEScalar0To1();
        return severity;
    }

}

File: src/cpp/cdm/patient/actions/SETensionPneumothorax.cpp
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

#include "cdm/CommonDefs.h"
#include "cdm/patient/actions/SETensionPneumothorax.h"
#include "cdm/properties/SEScalar0To1.h"
#include "cdm/io/protobuf/PBPatientActions.h"

SETensionPneumothorax::SETensionPneumothorax(Logger* logger) : SEPatientAction(logger)
{
  m_Type=eGate::NullGate;
  m_Side=eSide::NullSide;
  m_Severity=nullptr;
}

SETensionPneumothorax::~SETensionPneumothorax()
{
  m_Type = eGate::NullGate;
  m_Side = eSide::NullSide;
  SAFE_DELETE(m_Severity);
}

void SETensionPneumothorax::Clear()
{
  SEPatientAction::Clear();
  m_Type= eGate::NullGate;
  m_Side= eSide::NullSide;
  INVALIDATE_PROPERTY(m_Severity);
}

void SETensionPneumothorax::Copy(const SETensionPneumothorax& src, bool /*preserveState*/)
{
  //if(preserveState) // Cache any state before copy,
  PBPatientAction::Copy(src, *this);
  //if(preserveState) // Put back any state
}

bool SETensionPneumothorax::IsValid() const
{
  return SEPatientAction::IsValid() && HasType() && HasSide() && HasSeverity();
}

bool SETensionPneumothorax::IsActive() const
{
  if (!SEPatientAction::IsActive())
    return false;
  return !m_Severity->IsZero();
}
void SETensionPneumothorax::Deactivate()
{
  SEPatientAction::Deactivate();
  Clear();//No stateful properties
}

const SEScalar* SETensionPneumothorax::GetScalar(const std::string& name)
{
  if (name.compare("Severity") == 0)
    return &GetSeverity();
  return nullptr;
}

eGate SETensionPneumothorax::GetType() const
{
  return m_Type;
}
void SETensionPneumothorax::SetType(eGate Type)
{
  m_Type = Type;
}
bool SETensionPneumothorax::HasType() const
{
  return m_Type== eGate::NullGate ?false:true;
}
void SETensionPneumothorax::InvalidateType()
{
  m_Type = eGate::NullGate;
}

eSide SETensionPneumothorax::GetSide() const
{
  return m_Side;
}
void SETensionPneumothorax::SetSide(eSide Side)
{
  m_Side = Side;
}
bool SETensionPneumothorax::HasSide() const
{
  return m_Side== eSide::NullSide ?false:true;
}
void SETensionPneumothorax::InvalidateSide()
{
  m_Side = eSide::NullSide;
}

bool SETensionPneumothorax::HasSeverity() const
{
  return m_Severity==nullptr?false:m_Severity->IsValid();
}
SEScalar0To1& SETensionPneumothorax::GetSeverity()
{
  if(m_Severity==nullptr)
    m_Severity=new SEScalar0To1();
  return *m_Severity;
}
double SETensionPneumothorax::GetSeverity() const
{
  if (m_Severity == nullptr)
    return SEScalar::dNaN();
  return m_Severity->GetValue();
}
